esp-hal-common/src/i2c.rs

```diff
@@ -0,0 +1,968 @@
+            Command::Start | Command::Stop | Command::Read { .. } => Ack::NACK,
+            Command::Write { ack_exp: exp, .. } => exp,
+        };
+
+        let ack_check_en = match c {
+            Command::Start | Command::Stop | Command::Read { .. } => false,
+            Command::Write {
+                ack_check_en: en, ..
+            } => en,
+        };
+
+        let ack_value = match c {
+            Command::Start | Command::Stop | Command::Write { .. } => Ack::NACK,
+            Command::Read { ack_value: ack, .. } => ack,
+        };
+
+        let mut cmd: u16 = length.into();
+
+        if ack_check_en {
+            cmd |= 1 << 8;
+        } else {
+            cmd &= !(1 << 8);
+        }
+
+        if ack_exp == Ack::NACK {
+            cmd |= 1 << 9;
+        } else {
+            cmd &= !(1 << 9);
+        }
+
+        if ack_value == Ack::NACK {
+            cmd |= 1 << 10;
+        } else {
+            cmd &= !(1 << 10);
+        }
+
+        cmd |= (opcode as u16) << 11;
+
+        cmd
+    }
+}
+
+enum OperationType {
+    WRITE = 0,
+    READ  = 1,
+}
+
+#[derive(Eq, PartialEq, Copy, Clone)]
+enum Ack {
+    ACK,
+    NACK,
+}
+
+#[allow(dead_code)]
+#[cfg(any(feature = "esp32c3", feature = "esp32s3"))]
+enum Opcode {
+    RSTART = 6,
+    WRITE  = 1,
+    READ   = 3,
+    STOP   = 2,
+    END    = 4,
+}
+
+#[allow(dead_code)]
+#[cfg(any(feature = "esp32", feature = "esp32s2"))]
+enum Opcode {
+    RSTART = 0,
+    WRITE  = 1,
+    READ   = 2,
+    STOP   = 3,
+    END    = 4,
+}
+
+/// I2C peripheral container (I2C)
+pub struct I2C<T> {
+    peripheral: T,
+}
+
+impl<T> Read for I2C<T>
+where
+    T: Instance,
+{
+    type Error = Error;
+
+    fn read(&mut self, address: u8, buffer: &mut [u8]) -> Result<(), Self::Error> {
+        self.peripheral.master_read(address, buffer)
+    }
+}
+
+impl<T> Write for I2C<T>
+where
+    T: Instance,
+{
+    type Error = Error;
+
+    fn write(&mut self, addr: u8, bytes: &[u8]) -> Result<(), Self::Error> {
+        self.peripheral.master_write(addr, bytes)
+    }
+}
+
+impl<T> WriteRead for I2C<T>
+where
+    T: Instance,
+{
+    type Error = Error;
+
+    fn write_read(
+        &mut self,
+        address: u8,
+        bytes: &[u8],
+        buffer: &mut [u8],
+    ) -> Result<(), Self::Error> {
+        self.peripheral.master_write_read(address, bytes, buffer)
+    }
+}
+
+#[cfg(feature = "esp32")]
+type System = crate::pac::DPORT;
+#[cfg(not(feature = "esp32"))]
+type System = crate::pac::SYSTEM;
+
+impl<T> I2C<T>
+where
+    T: Instance,
+{
+    /// Create a new I2C instance
+    /// This will enable the peripheral but the periphal won't get automatically
+    /// disabled when this gets dropped.
+    pub fn new<
+        SDA: OutputPin<OutputSignal = OutputSignal> + InputPin<InputSignal = InputSignal>,
+        SCL: OutputPin<OutputSignal = OutputSignal> + InputPin<InputSignal = InputSignal>,
+    >(
+        i2c: T,
+        mut pins: Pins<SDA, SCL>,
```

Comment:
If any two GPIOs can be used for I2C, is there any real benefit to having this `Pins` type? Is there a reason the pins couldn't or shouldn't just be passed directly to the constructor?

Reply:
Probably just as a workaround for not having named parameters but I agree it's not really needed - will remove it